### DISCUSSION TITLE
arm64 packages for Ubuntu

### DISCUSSION THREAD

**Discussion by balhoff** [souffle-lang/souffle#2065] [General]
Would it be possible to provide pre-built Ubuntu packages for arm64? Our use case is including Soufflé in a Docker image we use on Apple M1 systems.

**Comment by XiaowenHu96** [comment]
We could but cross-compilation can be tricky, especially when we can't test the build on our side.
Alternatively, you can choose to build Souffle during the Docker setup.

**Reply by matentzn** [reply]
I would also like arm64 pre-built packages, if at all possible. It is a bit inconvenient to set up a simple docker pipeline to include software builds. You will probably need multistage builds, you need to uninstall build tools etc, etc.. But yes, in the worst case, we can build the packages ourselves.

**Reply by b-scholz** [reply]
Unfortunately, I don't have an Apple computer with an M1 chip. However, I tried to compile Souffle on a Raspberry Pi 400.  
It works mostly. I noticed that I had to add the `-latomic` link flag and adjust other bits and pieces of the build system. 

Also, I noticed that the Brie data structure has problems, and the interfaces tests use most likely the wrong data types. Overall, it would not be too much work, but I would need volunteers because we don't have the necessary hardware for the port. 

I

**Reply by XiaowenHu96** [reply]
I looked into how other maintainers deal with this issue.  Many of them do not release aarch64 simply because Github action doesn't provide an aarch64 runer. But it looks like there is some workaround by using docker. https://github.com/actions/virtual-environments/issues/2552#issuecomment-772404495

I think I can give it a try.

**Reply by b-scholz** [reply]
I don’t think an emulator is a good choice. The tests would take too long.

fingers crossed that GitHub will provide arm images in near future.

**Reply by balhoff** [reply]
Thanks for considering it! In the meantime we are looking into building Soufflé ourselves. Keeping all the tools to build Soufflé in our image makes it too large, so we're trying a multistage build.

**Reply by XiaowenHu96** [reply]
@balhoff
If you have access to an arm64, you can also build the package, upload it somewhere, then download and install it in your image. The command is 
```
cmake -S . -B ./build
cmake --build ./build -j --target package
```
More details here: https://github.com/souffle-lang/souffle/wiki/Packaging

**Reply by gouttegd** [reply]
Thanks @XiaowenHu96 for the precise packaging instructions, we'll try that.

For your information, on Ubuntu 20.04 on a M1 machine, the build goes smoothly and all tests pass except the following two:

```
The following tests FAILED:
        489 - evaluation/indexed_inequalities_compare_csv (Failed)
        1807 - semantic/negative_numbers_compare_csv (Failed)
```

Relevant sections of the `LastTest.log` file:

```
489/3061 Testing: evaluation/indexed_inequalities_compare_csv
489/3061 Test: evaluation/indexed_inequalities_compare_csv
Command: "/tools/souffle-2.1/cmake/check_test_results.sh" "/tools/souffle-2.1/tests/
evaluation/indexed_inequalities" ""
Directory: /tools/souffle-2.1/build2/tests/evaluation/indexed_inequalities_interpret
ed
"evaluation/indexed_inequalities_compare_csv" start time: Sep 17 09:32 UTC
Output:
----------------------------------------------------------
1d0
< 4294967290
5a5
> 4294967295
<end of output>
Test time =   0.01 sec
----------------------------------------------------------
Test Failed.
"evaluation/indexed_inequalities_compare_csv" end time: Sep 17 09:32 UTC
"evaluation/indexed_inequalities_compare_csv" time elapsed: 00:00:00
----------------------------------------------------------
[...]
1807/3061 Testing: semantic/negative_numbers_compare_csv
1807/3061 Test: semantic/negative_numbers_compare_csv
Command: "/tools/souffle-2.1/cmake/check_test_results.sh" "/tools/souffle-2.1/tests/
semantic/negative_numbers" ""
Directory: /tools/souffle-2.1/build2/tests/semantic/negative_numbers_interpreted
"semantic/negative_numbers_compare_csv" start time: Sep 17 09:53 UTC
Output:
----------------------------------------------------------
1d0
< -2147483647
<end of output>
Test time =   0.00 sec
----------------------------------------------------------
Test Failed.
"semantic/negative_numbers_compare_csv" end time: Sep 17 09:53 UTC
"semantic/negative_numbers_compare_csv" time elapsed: 00:00:00
----------------------------------------------------------
```

**Comment by b-scholz** [comment]
This test is failing:
```
.decl rel_2(x:unsigned)
.output rel_2

rel_2(2^32-5).
rel_2(x+1) :- rel_2(x), x >= 2^32-5, x < 2^32-1.
```

**Comment by b-scholz** [comment]
and the other one is tests/semantics/negative_numbers and is also related to negative numbers. 

It seems that the synthesised version is working - however the interpreted version fails. Most likely it is related to casts / and indexes in the interpreter.

**Comment by remysucre** [comment]
I can confirm the 2 failing tests above pass in 64-bit mode: `cmake -S . -B build -DSOUFFLE_DOMAIN_64BIT=ON`. This suggests they may be fixed by addressing the regression on #2171 (which was fixed by #2172 but now broken again, possibly by a9a6587).

**Reply by remysucre** [reply]
Indeed, all tests pass after reverting to #2172
